skimage/transform/_thin_plate_splines.py

```diff
@@ -0,0 +1,302 @@
+
+    Returns
+    -------
+    coeffs : ndarray
+        Array of shape (N+3, 2) containing the calculated coefficients.
+
+    """
+    n_coords = src.shape[1]
+    Y = np.row_stack((dst, np.zeros((n_coords+1, n_coords))))
+    L = _make_L_matrix(src)
+    coeffs = np.dot(np.linalg.pinv(L), Y)
+    return coeffs
+
+
+def _calculate_f(x, y, points, coeffs):
+    """Compute the thin-plate spline function at given coordinates (x, y).
+
+    Parameters:
+    ----------
+    coeffs : ndarray
+        Array of shape (N+3, 2) containing the thin-plate spline coefficients.
+    points : ndarray
+        Array of shape (N, 2) representing the source point.
+    x : ndarray
+        Array representing the x-coordinates of points to transform.
+    y : ndarray
+        Array representing the y-coordinates of points to transform.
+
+    Returns:
+    -------
+    ndarray :
+        Array containing the computed thin-plate spline function values.
+
+    Notes:
+    -----
+    This function calculates the thin-plate spline function values at each
+    coordinate (x, y) everywhere in the plane.
+
+    The function is calculated as:
+
+    .. math::
+
+        f(x, y) = a1 + ax * x + ay * y + Σ(wi * U(|Pi - (x, y)|))
+
+        where:
+        - a1, ax, and ay are the last three coefficients in `coeffs`.
+        - wi represents the weights corresponding to each point Pi in `points`.
+        - U(r) is the basis kernel function.
+
+    """
+    w = coeffs[:-3]
+    a1, ax, ay = coeffs[-3:]
+    summation = np.zeros(x.shape)
+    for wi, Pi in zip(w, points):
+        summation += wi * _U(np.sqrt((Pi[0]-x)**2 + (Pi[1]-y)**2))
+    return a1 + ax*x + ay*y + summation
+
+
+def tps_transform(src, dst, x_vals, y_vals):
+    """Apply transformation to coordinates `x_vals` and `y_vals`.
+
+    Parameters
+    ----------
+    src : (N, 2) array_like
+        An array of N points representing the source point.
+    dst : (N, 2) array_like
+        An array of N points representing the destination point.
+        `dst` must have the same shape as `src`.
+    x_vals : array_like
+        The x-coordinates of points to transform.
+    y_vals : array_like
+        The y-coordinates of points to transform.
+
+    Returns
+    -------
+    transformed_pts : lists
+        A list of transformed coordinates.
+
+    Examples
+    --------
+    >>> import skimage as ski
+
+    Define source and destination points:
+
+    >>> src = np.array([[0, 0], [0, 512], [512, 512],[512, 0]])
+    >>> dst = np.roll(src, 1, axis=0)
+
+    Generate the grid points for transformation
+
+    >>> x_min, x_max = 0, 90
+    >>> y_min, y_max = 0, 90
+    >>> num_points = 100
+    >>> x, y = np.meshgrid(np.linspace(x_min, x_max, num_points),
+    ...     np.linspace(y_min, y_max, num_points))
+
+    Apply the transformation
+
+    >>> x_trans, y_trans = ski.transform.tps_transform(src, dst, x, y)
+    >>> x_trans[10, 10]
+    501.88594868313635
+    >>> y_trans[10, 10]
+    10.112359869775215
```

Comment:
This result seems unstable depending on the precise environment (see [this failing CI job](https://github.com/scikit-image/scikit-image/actions/runs/5407340296/jobs/9825241253?pr=7040#step:5:539). But I think we can get a stable result if we simplify the test a little bit:

```suggestion
    Define source and destination points and generate meshgrid for transformation:
    >>> src = np.array([[0, 0], [0, 5], [5, 5],[5, 0]])
    >>> dst = np.roll(src, 1, axis=0)
    >>> xx, yy = np.meshgrid(np.arange(5), np.arange(5))
    >>> xx
    array([[0, 1, 2, 3, 4],
           [0, 1, 2, 3, 4],
           [0, 1, 2, 3, 4],
           [0, 1, 2, 3, 4],
           [0, 1, 2, 3, 4]])

    Apply the transformation

    >>> import skimage as ski
    >>> xx_trans, yy_trans = ski.transform.tps_transform(src, dst, x, y)
    >>> xx_trans
    array([[5., 5., 5., 5., 5.],
           [4., 4., 4., 4., 4.],
           [3., 3., 3., 3., 3.],
           [2., 2., 2., 2., 2.],
           [1., 1., 1., 1., 1.]])
```

what is also cool that this already shows the pixel-offset. Notice how `xx` goes from [0, 4], while `xx_trans` goes from [1.,5.]. :thinking:

Reply:
Yes, the offset is clear to see from this example. 
This would make a good test case.